### ISSUE TITLE
Build process is stuck at Python test

### ISSUE THREAD

**Issue by gzagatti** [nix-community/home-manager#4977]
I have been using nix and home-manager to manage the packages in my computer for the last two years. I adopted nix as replacement for brew. I have enjoyed the experience so far. The home-manager documentation is quite helpful to get one started. However, I have not had much time to learn nix and home-manager in much depth, so I find it difficult to debug the system.

I follow a very simple routine to update my packages with home-manager.

```sh
nix-channel --update
home-manager switch
```

However, recently I have faced an issue which I cannot get around. During the build phase, nix gets stuck at a Python test:

```
============================= test session starts ==============================
platform linux -- Python 3.10.13, pytest-7.4.3, pluggy-1.3.0
Test environment for tests-3726

System paths:
    sys.executable: /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/bin/python3.10
    sys.prefix: /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13
    sys.base_prefix: /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13
    sys.real_prefix: <missing>
    site.getsitepackages(): /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/lib/python3.10/site-packages
    site.getusersitepackages(): /homeless-shelter/.local/lib/python3.10/site-packages
    sys.path (site-packages): /nix/store/0fzprkr5g1sr0k835h503zfrw016p8x2-python3.10-werkzeug-2.3.8/lib/python3.10/site-packages
                              /nix/store/151pjjpyhiglqcd6nv31yy0c9yw1144y-python3.10-typing-extensions-4.8.0/lib/python3.10/site-packages
                              /nix/store/310kx41wy7hinnl4zykyw93qfknpz80m-python3.10-click-8.1.7/lib/python3.10/site-packages
                              /nix/store/4g6q6cr5pbz1hi4rbf0i7pzi2lrj8f5a-python3.10-iniconfig-2.0.0/lib/python3.10/site-packages
                              /nix/store/8fgcbhpwc54339dvh1xb4m2gkiv93d7n-python3.10-wheel-0.42.0/lib/python3.10/site-packages
                              /nix/store/a0kifv8g0hp8zvbydlz8mwqkpw9ckigk-python3.10-asgiref-3.7.2/lib/python3.10/site-packages
                              /nix/store/b001clk9jp6yl502fji8bqddx8m8159b-python3.10-requests-2.31.0/lib/python3.10/site-packages
                              /nix/store/b32sc8pjmdnkd5y3gc82rm4bf6sxyl3y-python3.10-zipp-3.17.0/lib/python3.10/site-packages
                              /nix/store/bl6kb0r1gfzh5ryq31ncc6mvd96bqvck-python3.10-pytest-timeout-2.2.0/lib/python3.10/site-packages
                              /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/lib/python3.10/site-packages
                              /nix/store/fal0ysy0anh3m3r5y5qlki9kbmqlgfh3-python3.10-setuptools-69.0.2/lib/python3.10/site-packages
                              /nix/store/g1ia17mcp79fzqq8iamm07x5z1aqlrzy-python3.10-debugpy-1.8.0/lib/python3.10/site-packages
                              /nix/store/gg918finnxkl08igs0pxi2y70wk9186a-python3.10-exceptiongroup-1.2.0/lib/python3.10/site-packages
                              /nix/store/h5l1m87a1bmbpaciyv0i448afcrn9y2n-python3.10-blinker-1.7.0/lib/python3.10/site-packages
                              /nix/store/h8sqsm2qzan80zh5dsq60pm9wx827ydk-python3.10-psutil-5.9.6/lib/python3.10/site-packages
                              /nix/store/hdb2rlw6rabjm78jz1ab3lycpa2aw0zb-python3.10-markupsafe-2.1.3/lib/python3.10/site-packages
                              /nix/store/i2a9bpzbxddhxhi5xq1f1frwq6314qx3-python3.10-jinja2-3.1.3/lib/python3.10/site-packages
                              /nix/store/kgj1dwqzs783wv85l8kagcz3v09s12yh-python3.10-Django-4.2.9/lib/python3.10/site-packages
                              /nix/store/kspxmhdgk28j4qd1vmkmk15k5g5rv04s-python3.10-idna-3.6/lib/python3.10/site-packages
                              /nix/store/kzyck6qg58xykmw7jy25zdz6s27d4fxw-python3.10-urllib3-2.1.0/lib/python3.10/site-packages
                              /nix/store/lcn0m613f5s7g2b62ww3a92vd6xy6hwl-python3.10-itsdangerous-2.1.2/lib/python3.10/site-packages
                              /nix/store/nlqr5vjaspnhgwma33iyjsi8xlsb52mi-python3.10-greenlet-3.0.1/lib/python3.10/site-packages
                              /nix/store/nwap2rsg2m0c3077hin9595xx709shwr-python3.10-tomli-2.0.1/lib/python3.10/site-packages
                              /nix/store/pfiwm6xywabyc2mrxfc9xdp1glsi201f-python3.10-cffi-1.16.0/lib/python3.10/site-packages
                              /nix/store/pgd2p6v4rs2h9gvhklmnl9p23fv8gx8r-python3.10-installer-0.7.0/lib/python3.10/site-packages
                              /nix/store/pq4b98aii196b2wp6i7lw1ixwn0nwsq7-python3.10-pluggy-1.3.0/lib/python3.10/site-packages
                              /nix/store/q4j9hzvizd28m5ywhnr2m1i038a53qri-python3.10-pycparser-2.21/lib/python3.10/site-packages
                              /nix/store/qq2fy1lk724z4baw6k8hwavl9xdl1bch-python3.10-zope.interface-5.5.2/lib/python3.10/site-packages
                              /nix/store/r7zmd3bb6a6vbs5i5nl8vs9z53zhhydg-python3.10-pytest-7.4.3/lib/python3.10/site-packages
                              /nix/store/ridvn1p20nz75145251v6qab2qc9gjqy-python3.10-importlib-metadata-6.9.0/lib/python3.10/site-packages
                              /nix/store/rn89wf8bg2rcq200cjg9ffh1fabgwfjw-python3.10-toml-0.10.2/lib/python3.10/site-packages
                              /nix/store/s4mcasqckzqhfgdlr11hn18wydzp8kg2-python3.10-flask-2.3.3/lib/python3.10/site-packages
                              /nix/store/s5sjxk8j4gxixwlwc89j0m50midicnar-python3.10-sqlparse-0.4.4/lib/python3.10/site-packages
                              /nix/store/smawnc62hy956y7xf5lkj4asfaiqs86n-python3.10-brotlicffi-1.1.0.0/lib/python3.10/site-packages
                              /nix/store/v79cljrmiwybrz2cvbmrkryw8yr1j9k3-python3.10-zope.event-4.6/lib/python3.10/site-packages
                              /nix/store/vi7vr1iajv6zdbdw02xra0ryzz1hbkdi-python3.10-charset-normalizer-3.3.2/lib/python3.10/site-packages
                              /nix/store/vqln5h860faa23pvsfw63vh29nindyis-python3.10-pytest-xdist-3.5.0/lib/python3.10/site-packages
                              /nix/store/w16zfyin9a129ck9cz5jp7gp96pd9dqx-python3.10-packaging-23.2/lib/python3.10/site-packages
                              /nix/store/xl32y250d861g4qq104p93271vwcp7hs-python3.10-certifi-2023.11.17/lib/python3.10/site-packages
                              /nix/store/yh1z6cn76sd4h0hs0cakrjzwk6n0n05b-python3.10-execnet-2.0.2/lib/python3.10/site-packages
                              /nix/store/z00660d6455xhagpkd5rgf5h88gqkixi-python3.10-gevent-23.9.1/lib/python3.10/site-packages
    sysconfig.get_path('stdlib'): /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/lib/python3.10
    sysconfig.get_path('platstdlib'): /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/lib/python3.10
    sysconfig.get_path('purelib'): /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/lib/python3.10/site-packages
    sysconfig.get_path('platlib'): /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/lib/python3.10/site-packages
    sysconfig.get_path('include'): /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/include/python3.10
    sysconfig.get_path('scripts'): /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/bin
    os.__file__: /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/lib/python3.10/os.py
    sysconfig.get_path('data'): /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13
    threading.__file__: /nix/store/cawqhcxl9ar46yws7jvscn64rn052kg2-python3-3.10.13/lib/python3.10/threading.py
    debugpy.__file__: /nix/store/g1ia17mcp79fzqq8iamm07x5z1aqlrzy-python3.10-debugpy-1.8.0/lib/python3.10/site-packages/debugpy/__init__.py

Installed packages:
    debugpy==1.8.0
    setuptools==69.0.2.post0
    wheel==0.42.0
    installer==0.7.0
    Django==4.2.9
    asgiref==3.7.2
    typing_extensions==4.8.0
    sqlparse==0.4.4
    Flask==2.3.3
    click==8.1.7
    blinker==1.7.0
    itsdangerous==2.1.2
    Jinja2==3.1.3
    MarkupSafe==2.1.3
    Werkzeug==2.3.8
    gevent==23.9.1
    importlib-metadata==6.9.0
    toml==0.10.2
    zipp==3.17.0
    zope.event==4.6
    zope.interface==5.5.2
    greenlet==3.0.1
    psutil==5.9.6
    pytest-timeout==2.2.0
    pytest-xdist==3.5.0
    execnet==2.0.2
    pytest==7.4.3
    iniconfig==2.0.0
    packaging==23.2
    pluggy==1.3.0
    exceptiongroup==1.2.0
    tomli==2.0.1
    requests==2.31.0
    brotlicffi==1.1.0.0
    cffi==1.16.0
    pycparser==2.21
    certifi==2023.11.17
    charset-normalizer==3.3.2
    idna==3.6
    urllib3==2.1.0
rootdir: /build/source
configfile: pytest.ini
testpaths: tests
plugins: timeout-2.2.0, xdist-3.5.0
8 workers [1216 items] m
........................................................................ [  5%]
........................................................................ [ 11%]
........................................................................ [ 17%]
```

The test does not conclude, nor the build times out. This has never happened before and is completely new to me. I am not even sure why we the formula needs to run these tests if we are only building packages.

Would anyone be able to help with possible workarounds? I am only raising this issue here as I am using nix exclusively with home-manager. However, please let me know if I should raise this issue elsewhere. As I mentioned before I have not learnt nix in much depth, so I do not know how to start debugging this problem. I am happy to share my configuration file or more information as required.

**Comment by Artturin** [comment]
The packages are from nixpkgs thus you should create an issue there but first search if there's one already.

**Comment by gzagatti** [comment]
Thanks, I will follow up in that repo. 

Do you have any suggestion on how to identify the offending package in my home-manager config file? I can only thing of a brute-force approach where I comment all my packages out and then activate them one by one. I tried to parse the logs of the build process but they are quite verbose, and I couldn't locate what I was looking for.

**Comment by gzagatti** [comment]
Thanks to @saaperezru I found the offending package. I would never have figured out without him opening that issue. 

I had installed jupyterlab with python 3.10 and had no idea that nix support had moved to 3.11. However, I find a bit of an overkill to specify which python version I need, when at the end of the day I just want a package that was written in python. Does anyone have any workaround to always pick the python package with the version currently supported by nixpkgs?

Thanks to @Artturin for mentioning the new cli.
